### PR TITLE
Rethinking resources for a v2 of the stack

### PR DESCRIPTION
**Pull request by tommyphilyaw**
Started roughing a prototype for the V2 of resources in the stack per #13. 

`service-demo.ts` serves as quick test bed and the end result of all this will be moving this out to packages and rewriting the hello world to use. Until then we are using the `framework` folder to put this code as we work through this.

### CODE REVIEW

**Comment by tommyphilyaw** [comment]
@lxndrdagreat The service result is a loose translation of something I did in internal project. Purely using it as a jumping off point for us to beat up and change.

It has worked well in that system but doesn't mean it's 100% what we need for this one. This also prefaces catching the errors in the service and returning an error vs throwing. Again based on my experience in the other internal project it's not a bad thing but maybe worth exploring both options.

Also haven't made it to actually creating the routes yet just what I'm calling the `ServiceDefinition` at the moment. Names of course can change.

File: unthink-stack/src/server/framework/service-definition.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { ServiceResult } from './service-result';
+
+type HandlerMethod = 'GET' | 'PUT' | 'POST' | 'DELETE';
+
+export interface HandlerContext {
+  query?: Record<string, string>;
+  path?: Record<string, string>;
```
Comment (lxndrdagreat) [inline unthink-stack/src/server/framework/service-definition.ts]:
```suggestion
  queryParams?: Record<string, string>;
  pathParams?: Record<string, string>;
```

Just a thought. In Angular, the `Route` information names them this way. To me, it clarifies what they are. `path` by itself does not sound like parameter information, but instead sounds like the URL.

File: unthink-stack/src/server/framework/service-result.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  ERROR,
+  NOT_FOUND
+}
+
+export class ServiceResult<ValueModel = unknown, ErrorModel = ServiceError> {
+
+  private readonly state: ServiceResultState;
+
+  public value?: ValueModel;
+  public error?: ErrorModel;
```
Comment (lxndrdagreat) [inline unthink-stack/src/server/framework/service-result.ts]:
```suggestion
  public readonly value?: ValueModel;
  public readonly error?: ErrorModel;
```

I don't think this is necessary, but if these are only supposed to be set via the constructor we can make them `readonly`.

File: unthink-stack/src/server/framework/service-definition.ts (hunk 1 above)
Comment (tommyphilyaw) [reply unthink-stack/src/server/framework/service-definition.ts]: Yeah totally agree. I think that was copied from how express does it, which to your point is even confusing there when you see `path.

File: unthink-stack/src/server/framework/service-demo.ts [hunk 3]
```diff
@@ -61,23 +62,25 @@ const sd: ServiceDefinition = {
...
-    ...
-
-    }
-*/
   }
 };
 
+/****** For templates ******/
+
+const td: TemplateDefintion = {
```
Comment (tommyphilyaw) [inline unthink-stack/src/server/framework/service-demo.ts]: @lxndrdagreat loose idea of a template oriented version of this - it's strictly templates not API calls here hence the split. 

Going back to wanting to keep the concepts apart for now.

File: unthink-stack/src/server/framework/service-demo.ts [hunk 4]
```diff
@@ -61,23 +62,25 @@ const sd: ServiceDefinition = {
...
+      return TemplateResult.view('home');
+    },
+  },
+  '/login': {
+    // Virtually no need for a class here just can directly map request to a view
+    'GET': async (ctx: HandlerContext): Promise<TemplateResult> => {
+      return TemplateResult.redirect('www.google.com');
+    },
+  }
+}
```
Comment (tommyphilyaw) [inline unthink-stack/src/server/framework/service-demo.ts]: @lxndrdagreat loose idea of a template oriented version of this - it's strictly templates not API calls here hence the split. 

Going back to wanting to keep the concepts apart for now.

**Comment by tommyphilyaw** [comment]
@lxndrdagreat below is where I landed with my latest. Alot of the resource-definition.ts stuff would be tucked away in the packages that generate that and really just the lines below would be the interface for the devs. 

https://github.com/epandco/unthink-cli/blob/5ee8415e8ff5da64147e279d38364164180f1192/unthink-stack/src/server/framework/service-demo.ts#L27-L54

**Comment by tommyphilyaw** [comment]
@lxndrdagreat Ok latest push is a another variation of this definition style we are going for. Names obviously probably need a deeper discussion when we solidify it but was going for more semantic meanings in the stack like `data` or `view` with the room to expand for like `file` etc when defining the kind of route.

This also lets you mix and match the various kinds of routes within a resource so the dev can choose how to package it up but each `route` has a specific type as mentioned above at the moment `data` and `view`. 

Data will generate the route with `/api` prefix so for `/users/:id` on the data route it will translate to `/api/users/:id` and this can be overriden by a property on that route or turned off etc.

A route is given path like `/users` or `/users/:id` and within that `route` you define the get/put/post/delete methods and as stated above this is locked down to a specific type meaning that if you define a view route for `/users/:id` all method within it must satisfy the 'view' handler and return a template result. 

It would be a bit weird for something like `/users/:id` to mix between methods e.g a `get` as a view route and `put` as a data route. 

It also brings it back into parity with middleware. I haven't full fleshed out types for middleware yet just did unknown as place holder and haven't do the actual express route generation but should not have all the info needed to do so.

Let me know your thoughts. I will attempt do the express piece tomorrow to get a full functional setup and we can just see how this feels.

**Comment by tommyphilyaw** [comment]
@lxndrdagreat ok not feature complete, however a good gauge for the feel of these new changes. I updated server.ts and reimplemented the hello world resource using v2. 

I think this will be close if not the final API for creating resources:

https://github.com/epandco/unthink-cli/blob/7939d1d99230aac6f49413944c38d9a2f606ba00/unthink-stack/src/server/resource-definitions/hello-world-resource-definition.ts#L1-L12

The `resource` function at the moment doesn't do anything other than let us inline the resource definition with the `export default` and thus don't need an intermediate variable that you have to name. However, once I get into middleware I may need that function to help with generics potentially.

To get us back to feature parity I still need to tackle:
- Middleware.
- Nunjucks default context / base context (e.g how version was globally available in all templates).
- Rethink dependency injection potentially or eliminate it.
- Similarly with logging requests reimplement logging.
- Tackle a error handler for `view` routes which are html. So if a handler throws an exception what do we do with it. 
- Bust src/framework out into packages that the unthink-stack consumes. For this sort of thing I want packages vs copying the code out with the template. That way if we make tweaks to how generation works or find bugs we aren't fixing them all overt he place.

I do want to talk about the following points though before I knock out some of this:
- How do we want to handle nunjucks default context in this new version
- Error handling for views - e.g what do we render:
  - Thinking we may need to handle / render pages like the old for 401 / 400 / 500 etc but maybe not. Don't know if we used them. 
- Extensibility of this framework. Route types specifically feel like there is opportunity to make them more `open` inline with SOLID principles. 
  - We have `view` and `data` baked in.
  - I think we need at least `file` route type baked in for upload/download.
  - Would like projects to be able to define their own and register it with the framework somehow potentially.

**Comment by tommyphilyaw** [comment]
Implemented strongly typed middleware for express through the use of generics in the resource definitions and an `expressResource` function for defining resources in express. 

There might be away around the `expressResource` function but I'm ok with that. Middleware is highly specific to the web server so if we implement this for KOA for example we would have a `koaResource` function that strongly types the middelware for KOA. 

It doesn't change how you define anything simply adds typing to middleware. 

Hello world example currently has a complete examples with all levels of middleware in place and if you run it you will get output similar too:

```
[server] resource level: 0
[server] view route level 0
[server] view method level 0

 
Quick note this block happens because hello world resource is at the root path '/'
and in dev mode express serves up the public folder so all those assets technically
get the resource level middleware applied because it flows through that path.

In production public is served up through nginx so this block would not be present.
---
[server] resource level: 1
[server] resource level: 2
[server] resource level: 3
[server] resource level: 4
[server] resource level: 5
---

[server] resource level: 6
[server] data route level - first 6
[server] data route level - second 6
[server] data method level - first 6
[server] data method level - second 6
[server] data method level - third 6
```

So far it all seems to be piped correctly and fires in the right order. You can also short the pipeline by having one the middleware components either call `next(... some error)` or by simply sending a response and not calling next. In both cases the target handler is skipped as expected. 

Same behavior we have in V1 of the stack and generally how middleware works.

**Comment by tommyphilyaw** [comment]
@lxndrdagreat I think I have sorted out all the major pieces and can start breaking this out into packages. I will likely kill this PR and open a new one that uses those packages once done.

That said if we get sometime I would like to talk through this one item from my list above:

- Error handing in `view` routes.

Currently in V2 your only option is to return a TemplateResult or throw an error unlike the `data` routes. I do think `view` routes need the same error semantics as `data` routes and here are my thoughts currently:

- There should be default pages for 400 / 401 / 404 / 500 like V1
  - Should these be actual routes and error handlers redirect to them vs rendering in place on the current route? 
  - Would impact the point below about adding model/value for context.
- Error objects for `view` maybe take:
  - an optional template arg so you can render a custom template
  - a model/value so you can given an error page context

**Comment by lxndrdagreat** [comment]
> There should be default pages for 400 / 401 / 404 / 500

I agree that we should have default templates for these (I thought we already did?). However, I am personally not a huge fan of having a defined route for them... at least for the 404 page. I'm used to sites not redirecting and instead just showing you the 404 error.

> - an optional template arg so you can render a custom template
> - a model/value so you can given an error page context

I like both of these.

**Comment by tommyphilyaw** [comment]
We do have default templates in current version (what I was calling at as `like v1`). 

I'm ok with it not redirecting to a specific route. The only times I can think of redirecting would be useful would be a 401 and having a default login page to redirect too but this can be solved in a variety of ways with middleware so lets just table that for now. 

Cool. I will head in that direction then for `view` routes and lets see how it feels.

**Comment by tommyphilyaw** [comment]
@lxndrdagreat Ok I think my last commit puts us back in parity with V1 without file handling. That needs to be in there but will likely take a different form than V1 but I would rather move forward with this and backfill file handling. 

Maybe we can quickly review this soon. If you are good with what I have then I will split this out into two packages

- UnthinkFoundation
- UnthinkFoundationExpress

For now Nunjucks is bundled with `UnthinkFoundationExpress`. Once we get this out there I do want to explore KOA and I think then would look at decoupling the template rendering and putting it into its own package to be shared across both backends.

**Comment by tommyphilyaw** [comment]
Closing this PR in preference of #35. Will leave the feature branch laying around briefly to ensure all things are ported to packages that need to be.

**Pull request closed without merging**